### ISSUE TITLE
IllegalAccessError in AndroidDeviceProperties

### ISSUE THREAD

**Issue by TheKeeperOfPie** [SiftScience/sift-android#8]
Not sure what causes this, but we've hit several crashes from users. Unfortunately I don't have reproduction steps.

For reference, this has only occurred on Xiaomi devices for now, a combination of MI 5s and Redmi Note 3/4.

Our Fabric crash can be viewed here: http://crashes.to/s/9a2d9a7276c

```
Fatal Exception: java.lang.IllegalAccessError: Field 'org.apache.commons.lang3.builder.HashCodeBuilder.iTotal' is inaccessible to class 'com.sift.api.representations.AndroidDevicePropertiesJson' (declaration of 'com.sift.api.representations.AndroidDevicePropertiesJson' appears in /data/app/com.seatgeek.android-1/base.apk)
       at com.sift.api.representations.AndroidDevicePropertiesJson.hashCode(AndroidDevicePropertiesJson.java:14967)
       at org.apache.commons.lang3.builder.HashCodeBuilder.append(HashCodeBuilder.java:850)
       at com.sift.api.representations.MobileEventJson.hashCode(MobileEventJson.java:178)
       at java.util.Collections.secondaryHash(Collections.java:3427)
       at java.util.WeakHashMap.put(WeakHashMap.java:593)
       at org.apache.commons.lang3.builder.ToStringStyle.register(ToStringStyle.java:182)
       at org.apache.commons.lang3.builder.ToStringStyle.appendClassName(ToStringStyle.java:1444)
       at org.apache.commons.lang3.builder.ToStringStyle.appendStart(ToStringStyle.java:369)
       at org.apache.commons.lang3.builder.ToStringBuilder.<init>(ToStringBuilder.java:263)
       at org.apache.commons.lang3.builder.ReflectionToStringBuilder.<init>(ReflectionToStringBuilder.java:456)
       at org.apache.commons.lang3.builder.ReflectionToStringBuilder.toString(ReflectionToStringBuilder.java:282)
       at org.apache.commons.lang3.builder.ReflectionToStringBuilder.toString(ReflectionToStringBuilder.java:113)
       at org.apache.commons.lang3.builder.ToStringBuilder.reflectionToString(ToStringBuilder.java:152)
       at com.sift.api.representations.MobileEventJson.toString(MobileEventJson.java:173)
       at siftscience.android.Queue.append(Queue.java:199)
       at siftscience.android.Sift.collect(Sift.java:1090)
       at com.seatgeek.android.analytics.SiftScienceTrackerImpl.access$000(SiftScienceTrackerImpl.java:1100)
       at com.seatgeek.android.analytics.SiftScienceTrackerImpl$1.onActivityCreated(SiftScienceTrackerImpl.java:47)
       at android.app.Application.dispatchActivityCreated(Application.java:199)
       at android.app.Activity.onCreate(Activity.java:925)
       at android.support.v4.app.FragmentActivity.onCreate(FragmentActivity.java:278)
       at android.support.v7.app.AppCompatActivity.onCreate(AppCompatActivity.java:85)
       at com.trello.rxlifecycle.components.support.RxAppCompatActivity.onCreate(RxAppCompatActivity.java:45)
       at com.seatgeek.android.ui.activities.BaseFragmentActivity.onCreate(BaseFragmentActivity.java:265)
       at com.seatgeek.android.ui.activities.SplashActivity.onCreate(SplashActivity.java)
       at android.app.Activity.performCreate(Activity.java:6288)
       at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1113)
       at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2500)
       at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2613)
       at android.app.ActivityThread.access$900(ActivityThread.java:180)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1473)
       at android.os.Handler.dispatchMessage(Handler.java:111)
       at android.os.Looper.loop(Looper.java:207)
       at android.app.ActivityThread.main(ActivityThread.java:5710)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:900)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:761)
```

**Comment by quanpod** [comment]
Thanks for the report @TheKeeperOfPie - please let us know if you're able to find a repro.

**Comment by TheKeeperOfPie** [comment]
Has there been any progress here? We've had dozens of more crashes and It does seem to only be occurring on Xiaomi devices, but unfortunately we don't have one to test with.

Does Sift have access to one and could confirm the crash? Might just be an OEM quirk that has be coded around.

**Comment by quanpod** [comment]
Hi @TheKeeperOfPie - we don't have a device on hand, so we may need to order one to find a repro here.  Do you have an idea of how often it's occurring on Xiaomi devices (e.g. how often we'd have to run Seatgeek on this device to encounter this situation?)?  I'm still concerned that even with a device we may not be able to definitively get a reproducible bug that we can fix.  We also do have a new release of the Android SDK out that's worth updating to and watching to see if we still encounter issues here.

**Comment by TheKeeperOfPie** [comment]
Following up on this, we still see quite a few crashes from Xiaomi devices.

I got my hands on a Redmi Note 4, which is the most represented of the devices and unfortunately cannot reproduce the crash.

But it seems to related to the fact that `jsonschema2pojo-gradle-plugin` uses reflection in the `toString()` method of its models. Xiaomi devices may have some undefined behavior when you try reflection like that.

Perhaps there's alternative model generator which doesn't use reflection?

**Comment by quanpod** [comment]
Thanks @TheKeeperOfPie - @garylee1 will be the person taking a look here

**Comment by garylee1** [comment]
@TheKeeperOfPie this should be addressed by our upcoming release!

**Comment by garylee1** [comment]
this has been fixed in 0.9.9

**Comment by garylee1** [comment]
please hold off – we're fixing a small issue related to deserializing archived configs. the fix will be released soon with 0.9.10.

**Comment by garylee1** [comment]
0.9.10 has been released – please skip over 0.9.9
